src/Apropos/Gen/BacktrackingTraversal.hs

```diff
@@ -0,0 +1,125 @@
+module Apropos.Gen.BacktrackingTraversal (
+  Traversal (..),
+  traversalAsGen,
+  traversalContainRetry,
+) where
+
+import Apropos.Gen
+import Apropos.HasPermutationGenerator.Morphism
+import Control.Monad ((>=>))
+import Control.Monad.Trans.Except (runExceptT)
+import Control.Monad.Trans.Writer (runWriterT)
+import Data.String (fromString)
+import Hedgehog (PropertyT)
+import Hedgehog qualified as H
+
+data Traversal p m where
+  Source :: Gen m -> Traversal p m
+  Traversal :: Traversal p m -> (m -> Gen [Morphism p m]) -> Traversal p m
+
+traversalAsGen :: Traversal p m -> Gen m
+traversalAsGen (Source g) = g
+traversalAsGen (Traversal m g) = do
+  n <- traversalAsGen m
+  t <- g n
+  foldl (>=>) pure (morphism <$> t) n
+
+traversalContainRetry :: Show m => Int -> Traversal p m -> PropertyT IO m
+traversalContainRetry retries m = do
+  r <- traversalWithRetries retries m
+  case r of
+    Nothing -> H.footnote "retry limit reached" >> H.discard
+    Just so -> pure so
+
+traversalWithRetries :: forall p a. Show a => Int -> Traversal p a -> PropertyT IO (Maybe a)
+traversalWithRetries retries m =
+  backtrackingRetryTraversals retries (forAllRetryToMaybeScale s) (genTraversal <$> t)
+  where
+    st :: (Gen a, [a -> Gen [Morphism p a]])
+    st = unTraversal m
+    s :: Gen a
+    s = fst st
+    t :: [a -> Gen [Morphism p a]]
+    t = snd st
+    genTraversal :: (a -> Gen [Morphism p a]) -> (a -> PropertyT IO [Int -> a -> PropertyT IO (Maybe a)])
+    genTraversal gt = \a -> do
+      tr <- forAllWith (unwords . (name <$>)) $ prune (gt a)
+      pure $ sizableTr <$> (morphism <$> tr)
+    sizableTr :: (a -> Gen a) -> (Int -> a -> PropertyT IO (Maybe a))
+    sizableTr g = \si a -> forAllRetryToMaybeScale (g a) si
+
+unTraversal :: Traversal p a -> (Gen a, [a -> Gen [Morphism p a]])
+unTraversal (Source s) = (s, [])
+unTraversal (Traversal s t) = case unTraversal s of
+  (s', t') -> (s', t' <> [t])
+
+forAllRetryToMaybeScale :: Show a => Gen a -> Int -> PropertyT IO (Maybe a)
+forAllRetryToMaybeScale g s = do
+  (ee, labels) <- H.forAll $ runWriterT (runExceptT $ gen $ scale (2 * s +) g)
+  mapM_ (H.label . fromString) labels
+  case ee of
+    Left Retry -> pure Nothing
+    Left (GenException err) -> H.footnote err >> H.failure
+    Right a -> pure $ Just a
+
+backtrackingRetryTraversals ::
+  forall a m.
+  Monad m =>
+  Int ->
+  (Int -> m (Maybe a)) ->
+  [a -> m [Int -> a -> m (Maybe a)]] ->
+  m (Maybe a)
+backtrackingRetryTraversals retries g trs = go 1
+  where
+    go :: Int -> m (Maybe a)
+    go l = do
+      res <- g l
+      case res of
+        Nothing ->
+          if l > retries
+            then pure Nothing
+            else go (l + 1)
```

Comment:
There is a bit of repetition of stuff like this that could be refactored away. I am thinking it would be nice to have more control over retry limits on a per morphism/source basis. This would involve refactoring this logic so I suggest we leave the repetition in for now and get rid of it when we add these improvements.

Reply:
If we don't end up wanting to handle the two points of failure differently MaybeT is kinda nice for this.
```haskell
    go :: Int -> m (Maybe a)
    go l =
      runMaybeT
        ( do
            so <- MaybeT $ g l
            MaybeT $ co so 1 trs
        )
        >>= \case
          Just so' -> pure $ Just so'
          Nothing
            | l > retries -> pure Nothing
            | otherwise -> go (l + 1)
```

Reply:
This looks nice! I think I will leave this for another PR though :)